protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+		hash_lock: HashLock<Self::Hash>,
+		time_lock: TimeLock,
+		amount: Amount,
+	) -> BridgeContractInitiatorResult<()> {
+		let contract = AtomicBridgeInitiator::new(self.initiator_address.0, &self.rpc_provider);
+		let recipient_bytes: [u8; 32] = recipient_address.0.try_into().unwrap();
+		let call = contract.initiateBridgeTransfer(
+			U256::from(amount.0),
+			FixedBytes(recipient_bytes),
+			FixedBytes(hash_lock.0),
+			U256::from(time_lock.0),
+		);
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn complete_bridge_transfer(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+		pre_image: HashLockPreImage,
+	) -> BridgeContractInitiatorResult<()> {
+		let pre_image: [u8; 32] = utils::vec_to_array(pre_image.0)
+			.unwrap_or_else(|_| panic!("Failed to convert pre_image"));
+		let contract = AtomicBridgeInitiator::new(self.initiator_address.0, &self.rpc_provider);
+		let call = contract
+			.completeBridgeTransfer(FixedBytes(bridge_transfer_id.0), FixedBytes(pre_image));
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn refund_bridge_transfer(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+	) -> BridgeContractInitiatorResult<()> {
+		let contract = AtomicBridgeInitiator::new(self.initiator_address.0, &self.rpc_provider);
+		let call = contract.refundBridgeTransfer(FixedBytes(bridge_transfer_id.0));
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn get_bridge_transfer_details(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+	) -> BridgeContractInitiatorResult<Option<BridgeTransferDetails<Self::Address, Self::Hash>>> {
+		let mapping_slot = U256::from(0); // the mapping is the zeroth slot in the contract
+		let key = bridge_transfer_id.0;
+		let storage_slot = self.calculate_storage_slot(key, mapping_slot);
+		let storage: U256 = self
+			.rpc_provider
+			.get_storage_at(self.initiator_address.0, storage_slot)
+			.await
+			.unwrap_or_else(|_| panic!("Failed to get storage at slot"));
+		let storage_bytes = storage.to_be_bytes::<32>();
+		let mut storage_slice = &storage_bytes[..];
+		let eth_details = EthBridgeTransferDetails::decode(&mut storage_slice).unwrap();
+
+		let details = BridgeTransferDetails {
+			bridge_transfer_id,
+			initiator_address: InitiatorAddress(eth_details.originator),
+			recipient_address: RecipientAddress(eth_details.recipient.to_vec()),
+			hash_lock: HashLock(eth_details.hash_lock),
+			time_lock: TimeLock(eth_details.time_lock.wrapping_to::<u64>()),
+			amount: Amount(eth_details.amount.wrapping_to::<u64>()),
+		};
+
+		Ok(Some(details))
+	}
+}
```

Comment:
Interesting read, as I conclude when the smart contract changes the slot number change. Complicated stuff, just avoid the panics.